File: backend/src/main/kotlin/com/trackwatch/backend/clients/email/ResendClient.kt
package com.trackwatch.backend.clients.email

import com.resend.Resend
import com.resend.services.emails.model.CreateEmailOptions
import com.trackwatch.backend.exception.InternalServerErrorException
import com.trackwatch.backend.model.User
import com.trackwatch.backend.service.MetricService
import com.trackwatch.backend.exception.ErrorCode
import com.trackwatch.backend.utils.values.Metrics
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class ResendClient(private val metricService: MetricService) {

    private val log = LoggerFactory.getLogger(ResendClient::class.java)
    private val resend = Resend(System.getenv("RESEND_API_KEY"))

    fun sendEmail(recipient: User, emailSubject: String, emailBody: String) {
        val params = createEmailParams(recipient, emailSubject, emailBody)

        try {
            sendMetricEmail()
            resend.emails().send(params)

        } catch (e: Exception) {
            log.error("Failed to send email: ${e.message}")
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, "Failed to send email: ${e.message}")
        }
    }

    private fun createEmailParams(recipient: User, emailSubject: String, emailBody: String): CreateEmailOptions {
        return CreateEmailOptions.builder()
            .from("TrackWatch <trackwatch@emlopezr.com>")
            .to(recipient.email)
            .subject(emailSubject)
            .html(emailBody)
            .build()
    }

    private fun sendMetricEmail() {
        metricService.incrementCounter(
            Metrics.EMAIL_SENT,
            "provider", "resend"
        )
    }
}

File: backend/app/clients/email/resend_client.py
import resend
from app.constants import AppInfo
from decouple import config
from app.exceptions import InternalServerErrorException, ErrorCode
from app.models.user import User

def _configure_api_key():
  api_key = config("RESEND_API_KEY", default="")
  if not api_key: return False
  resend.api_key = api_key
  return True

def send_admin_email(admin_email, email_subject, email_body):
  if not _configure_api_key(): return
  params = create_email_params(admin_email, email_subject, email_body)

  try:
    resend.Emails.send(params)
  except Exception as e:
    print(f"Failed to send admin email: {str(e)}")

def send_email(recipient, email_subject, email_body):
  if not _configure_api_key(): return
  params = create_email_params(recipient, email_subject, email_body)

  try:
    resend.Emails.send(params)
  except Exception as e:
    print(f"Failed to send email: {str(e)}")
    raise InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, f"Failed to send email: {str(e)}")

def create_email_params(recipient, email_subject, email_body) -> resend.Emails.SendParams:
  email_from = generate_email_from(AppInfo.APP_NAME, AppInfo.DOMAIN)
  email_to = [recipient.email if isinstance(recipient, User) else recipient]

  return {
    "from": email_from,
    "to": email_to,
    "subject": email_subject,
    "html": email_body
  }

def generate_email_from(app_name, domain):
  subdomain = app_name.lower()
  return f"{app_name} <{subdomain}@{domain}>"
